libcst/codemod/commands/apply_stub_annotations.py

```diff
@@ -0,0 +1,506 @@
+        # Qualifier for storing the canonical name of the current function.
+        self.qualifier: List[str] = []
+        # Store the annotations.
+        self.function_annotations: Dict[str, FunctionAnnotation] = {}
+        self.attribute_annotations: Dict[str, cst.Annotation] = {}
+        self.imports: Dict[str, ImportStatement] = {}
+        self.existing_imports: Set[str] = existing_imports
+        self.class_definitions: Dict[str, cst.ClassDef] = {}
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> None:
+        self.qualifier.append(node.name.value)
+        self.class_definitions[node.name.value] = node
+
+    def leave_ClassDef(self, original_node: cst.ClassDef) -> None:
+        self.qualifier.pop()
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
+        self.qualifier.append(node.name.value)
+        returns = node.returns
+        if returns is not None:
+            return_annotation = self._create_import_from_annotation(returns)
+            parameter_annotations = self._import_parameter_annotations(node.params)
+            self.function_annotations[".".join(self.qualifier)] = FunctionAnnotation(
+                parameters=parameter_annotations, returns=return_annotation
+            )
+        # pyi files don't support inner functions, return False to stop the traversal.
+        return False
+
+    def leave_FunctionDef(self, original_node: cst.FunctionDef) -> None:
+        self.qualifier.pop()
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool:
+        # pyre-fixme[16]: `BaseExpression` has no attribute `value`.
+        self.qualifier.append(node.target.value)
+        annotation_value = self._create_import_from_annotation(node.annotation)
+        self.attribute_annotations[".".join(self.qualifier)] = annotation_value
+        return True
+
+    def leave_AnnAssign(self, original_node: cst.AnnAssign) -> None:
+        self.qualifier.pop()
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+        module = node.module
+        if module is None or isinstance(node, cst.ImportStar):
+            return
+        # pyre-fixme[6]: Expected `List[CSTNode]` for 1st param but got
+        #  `Union[Sequence[ImportAlias], ImportStar]`.
+        # pyre-fixme[6]: Expected `str` for 1st param but got `Union[BaseExpression,
+        #  str]`.
+        self._add_to_imports(node.names, cst.Name(module.value), module.value)
+
+    def _add_annotation_to_imports(
+        self, annotation: cst.Attribute
+    ) -> Union[cst.Name, cst.Attribute]:
+        key = _get_attribute_as_string(annotation.value)
+        # Don't attempt to re-import existing imports.
+        if key in self.existing_imports:
+            return annotation
+        self._add_to_imports(
+            [cst.ImportAlias(name=annotation.attr)], annotation.value, key
+        )
+        return annotation.attr
+
+    @overload
+    def _handle_Index(self, slice: cst.Index, node: cst.Subscript) -> cst.Subscript:
```

Comment:
I'm unfamiliar with this code, but I'm lost on the type hints here.  The function body makes assumptions about `node` (namely that it has a `.slice`) without enforcing that.  Is the overload necessary?